src/lib/support/Pool.h

```diff
@@ -0,0 +1,125 @@
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *   Defines a memory pool class BitMapObjectPool.
+ */
+
+#pragma once
+
+#include <array>
+#include <assert.h>
+#include <atomic>
+#include <limits>
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+class StaticAllocatorBase
+{
+public:
+    StaticAllocatorBase(size_t capacity) : mAllocated(0), mCapacity(capacity) {}
+
+    size_t Capacity(void) const { return mCapacity; }
+    size_t Allocated(void) const { return mAllocated; }
+    bool Exhausted(void) const { return mAllocated == mCapacity; }
+
+protected:
+    size_t mAllocated;
+    const size_t mCapacity;
+};
+
+class StaticAllocatorBitmap : public StaticAllocatorBase
+{
+protected:
+    /**
+     * Use the largest data type supported by `std::atomic`. Putting multiple atomic inside a single cache line won't improve
+     * concurrency, while the use of larger data type can improve the performance by reducing the number of outer loop iterations.
+     */
+    using tBitChunkType                         = unsigned long;
+    static constexpr const tBitChunkType kBit1  = 1; // make sure bitshifts produce the right type
+    static constexpr const size_t kBitChunkSize = std::numeric_limits<tBitChunkType>::digits;
+    static_assert(ATOMIC_LONG_LOCK_FREE, "StaticAllocatorBitmap is not lock free");
+
+public:
+    StaticAllocatorBitmap(void * storage, std::atomic<tBitChunkType> * usage, size_t capacity, size_t elementSize);
+    void * Allocate(void);
+    void Deallocate(void * element);
+
+    // Test-only function declaration
+    template <typename F>
+    void ForEachActiveObject(F f);
+
+private:
+    void * At(size_t index) { return static_cast<uint8_t *>(mElements) + mElementSize * index; }
+    size_t IndexOf(void * element) {
+        std::ptrdiff_t diff = static_cast<uint8_t *>(element) - static_cast<uint8_t *>(mElements);
+        assert(diff > 0);
+        assert(static_cast<size_t>(diff) % mElementSize == 0);
+        auto index = static_cast<size_t>(diff) / mElementSize;
+        assert(index < Capacity());
+        return index;
+    }
+
+private:
+    void * mElements;
+    const size_t mElementSize;
+    std::atomic<tBitChunkType> * mUsage;
+};
+
+/**
+ *  @brief
+ *   A class template used for allocating Objects.
+ *
+ *  @tparam     T   a subclass of element to be allocated.
+ *  @tparam     N   a positive integer max number of elements the pool provides.
+ */
+template <class T, size_t N>
+class BitMapObjectPool : public StaticAllocatorBitmap
+{
+public:
+    BitMapObjectPool() : StaticAllocatorBitmap(mMemory, mUsage, N, sizeof(T)) {}
+
+    static size_t Size() { return N; }
+
+    template <typename... Args>
+    T * New(Args &&... args)
+    {
+        T * element = reinterpret_cast<T *>(Allocate());
+        if (element != nullptr)
+            return new (element) T(std::forward<Args>(args)...);
+        else
+            return nullptr;
+    }
+
+    void Delete(T * element)
```

Comment:
Is it intentionally that this is not scoped overloads of  C++ global delete?

Reply:
No, it is not scoped overloads of new/delete. The interface is schematically different from new/delete, I'll change it to other names to avoid confusion.